UnityServer/functions/src/model/UnityManager.ts

```diff
@@ -0,0 +1,203 @@
+   if there is a priceChange field and the state of the field is 0 = pending.
+   */
+  async checkSubscriptionPriceChange(userId: string): Promise<UnityStatus> {
+    return this.usersCollectionReference.doc(userId).collection("subscription").doc("receipt").get()
+      .then(doc => {
+        if (doc.exists) {
+          const receipt = {
+            packageName: doc.data() ?.packageName,
+            productId: doc.data() ?.productId,
+            purchaseToken: doc.data() ?.purchaseToken
+          }
+
+          return this.verifier.verifySub(receipt).then((result: any) => {
+            console.log(result.payload);
+            if (result.payload.priceChange !== undefined && result.payload.priceChange.state === 0) {
+              return new UnityStatus(true, "Subscription price change and has not been accepted by user", receipt.productId);
+            } else {
+              return new UnityStatus(false, "Subscription price has not change or has been accepted by user");
+            }
+          }).catch((error: any) => {
+            return new UnityStatus(false, error);
+          });
+        } else {
+          return new UnityStatus(false, "No active subscription");
+        }
+      }).catch(error => {
+        return new UnityStatus(false, error);
+      });
+  }
+  /* This function will verify with the json receipt send from the Google Play Billing api
+     To the client for additional security.
+      1) It will verify that the receipt is legit with the help of Google Play Developer api.
+      2) And will verify if the purchaseToken is unique in the database.
+     * more information : https://developer.android.com/google/play/billing/security
+  */
+  async verifyAndSavePurchase(userId: string, jsonReceipt: string): Promise<UnityStatus> {
+    const purchaseInformation = this.getReceiptAndPurchaseType(jsonReceipt);
+
+    if (purchaseInformation.purchaseType === SkuType.ONE_TIME) {
+      return this.verifier.verifyINAPP(purchaseInformation.receipt).then(() => {
+        return this.verifyPurchaseAndSaveInDB(userId, purchaseInformation.receipt);
+      }).catch((error: any) => {
+        return new UnityStatus(false, error.errorMessage);
+      });
+    } else {
+      return this.verifier.verifySub(purchaseInformation.receipt).then(async () => {
+        const unityStatus = await this.verifyPurchaseAndSaveInDB(userId, purchaseInformation.receipt);
+        if (unityStatus.success) {
+          return this.saveSubscriptionTokenInUsers(userId, purchaseInformation.receipt);
+        }
+        return unityStatus;
+      }).catch((error: any) => {
+        return new UnityStatus(false, error.errorMessage);
+      });
+    }
+  }
+
+  // 2) verification of the purchase token in database and if it's not in the database
+  //    it will save the receipt in purchases table.
+  async verifyPurchaseAndSaveInDB(userId: string, receipt: any): Promise<UnityStatus> {
```

Comment:
Is .ts a script language? Does it have concept of private/public?

Reply:
Yes, to my understanding, if there's no modifier like 'private', it will be public.